libs/cache/file_cache.go

```diff
@@ -0,0 +1,269 @@
+
+func (fc *FileCache[T]) addTelemetryMetric(key string) {
+	if fc.metrics != nil {
+		fc.metrics.SetBoolValue(key, true)
+	}
+}
+
+// GetOrCompute retrieves cached content or computes it using the provided function.
+// Cache operations fail open: if caching fails, the compute function is still called.
+// When cacheEnabled is false, the cache checks if values exist and measures potential time savings,
+// but always computes and never returns cached values.
+func (fc *FileCache[T]) GetOrCompute(ctx context.Context, fingerprint any, compute func(ctx context.Context) (T, error)) (T, error) {
+	// Convert fingerprint to deterministic hash - this is our cache key
+	cacheKey, err := fingerprintToHash(fingerprint)
+	if err != nil {
+		// Fail open: if we can't generate cache key, just compute directly
+		log.Debugf(ctx, "[Local Cache] failed to generate cache key, computing without cache: %v\n", err)
+		return compute(ctx)
+	}
+
+	log.Debugf(ctx, "[Local Cache] using cache key: %s\n", cacheKey)
+	fc.addTelemetryMetric("local.cache.attempt")
+
+	cachePath := fc.getCachePath(cacheKey)
+
+	// Try to read from disk cache
+	cachedData, cacheExists := fc.readFromCache(ctx, cachePath)
+
+	// Record metrics
+	if cacheExists {
+		log.Debugf(ctx, "[Local Cache] cache hit\n")
+		fc.addTelemetryMetric("local.cache.hit")
+
+		// If cache is enabled, return the cached value
+		if fc.cacheEnabled {
+			return cachedData, nil
+		}
+	} else {
+		log.Debugf(ctx, "[Local Cache] cache miss, computing\n")
+		fc.addTelemetryMetric("local.cache.miss")
+	}
+
+	// Acquire lock to prevent concurrent computations and writes for the same cache key
+	fc.mu.Lock()
+	defer fc.mu.Unlock()
+
+	// Compute the value and measure timing
+	start := time.Now()
+	result, err := compute(ctx)
+	if err != nil {
+		log.Debugf(ctx, "[Local Cache] error while computing: %v\n", err)
+		fc.addTelemetryMetric("local.cache.error")
+		return result, err
+	}
+
+	// Record duration metrics
+	if fc.metrics != nil {
+		computeDuration := time.Since(start)
+		fc.metrics.SetDurationValue("local.cache.compute_duration", computeDuration)
+	}
+
+	log.Debugf(ctx, "[Local Cache] computed and stored result\n")
+
+	// Write to disk cache (failures are silent - cache write errors don't affect the result)
+	fc.writeToCache(ctx, cachePath, result)
+
+	return result, nil
+}
+
+// readFromCache attempts to read and deserialize data from the cache file.
+// Expiry is checked using file modification time for consistency with cleanup.
+func (fc *FileCache[T]) readFromCache(ctx context.Context, cachePath string) (T, bool) {
+	var zero T
+
+	// Check file modification time for expiry
+	info, err := os.Stat(cachePath)
+	if err != nil {
+		log.Debugf(ctx, "[Local Cache] failed to stat cache file: %v\n", err)
+		return zero, false
+	}
+
+	age := time.Since(info.ModTime())
+	expiryDuration := time.Duration(fc.expiryMinutes) * time.Minute
+	if age > expiryDuration {
```

Comment:
This is another case of computing the expiry time.

You could add a member function to return the expiry threshold time and do simple comparison with Before/After on that instead of reproducing the logic.